server/src/main/java/org/elasticsearch/search/diversification/ResultDiversificationRetrieverBuilder.java

```diff
@@ -0,0 +1,313 @@
+            rankWindowSize,
+            queryVector,
+            lambda
+        );
+    }
+
+    @Override
+    public ActionRequestValidationException validate(
+        SearchSourceBuilder source,
+        ActionRequestValidationException validationException,
+        boolean isScroll,
+        boolean allowPartialSearchResults
+    ) {
+        // ensure the type is one we know of - at the moment, only "mmr" is valid
+        if (diversificationType.equals(DIVERSIFICATION_TYPE_MMR) == false) {
+            validationException = addValidationError(
+                String.format(Locale.ROOT, "[%s] diversification type must be set to [%s]", getName(), DIVERSIFICATION_TYPE_MMR),
+                validationException
+            );
+        }
+
+        // if MMR, ensure we have a lambda between 0.0 and 1.0
+        if (diversificationType.equals(DIVERSIFICATION_TYPE_MMR) && (lambda == null || lambda < 0.0 || lambda > 1.0)) {
+            validationException = addValidationError(
+                String.format(
+                    Locale.ROOT,
+                    "[%s] MMR result diversification must have a [%s] between 0.0 and 1.0",
+                    getName(),
+                    LAMBDA_FIELD.getPreferredName()
+                ),
+                validationException
+            );
+        }
+
+        return validationException;
+    }
+
+    @Override
+    protected RetrieverBuilder doRewrite(QueryRewriteContext ctx) {
+        IndexVersion indexVersion = ctx.getIndexSettings().getIndexVersionCreated();
+        Mapper mapper = ctx.getMappingLookup().getMapper(diversificationField);
+        if (mapper instanceof DenseVectorFieldMapper == false) {
+            throw new IllegalArgumentException("[" + diversificationField + "] is not a dense vector field");
+        }
+
+        if (diversificationType.equals(DIVERSIFICATION_TYPE_MMR)) {
+            diversificationContext = new MMRResultDiversificationContext(
+                diversificationField,
+                lambda == null ? DEFAULT_LAMBDA_VALUE : lambda,
+                rankWindowSize,
+                (DenseVectorFieldMapper) mapper,
+                indexVersion,
+                queryVector == null ? null : new VectorData(queryVector),
+                null
+            );
+        } else {
+            // should not happen
+            throw new IllegalArgumentException("Unknown diversification type [" + diversificationType + "]");
+        }
+
+        return this;
+    }
+
+    @Override
+    protected SearchSourceBuilder finalizeSourceBuilder(SearchSourceBuilder sourceBuilder) {
+        return super.finalizeSourceBuilder(sourceBuilder).docValueField(diversificationField);
+    }
+
+    @Override
+    protected RankDoc[] combineInnerRetrieverResults(List<ScoreDoc[]> rankResults, boolean explain) {
+        // must have the combined result set
+        assert rankResults.size() == 1 : "ResultDiversificationRetrieverBuilder must have a single result set";
```

Comment:
Lots of asserts here, is there any way they could ever hit? Thinking of serverless on-call suppressed rest errors here. This would trigger an on-call response as it would be a failed assertion.